lib/ui/window.dart

```diff
@@ -143,14 +161,234 @@ class Locale {
   /// ```
   ///
   /// The primary language subtag must not be null. The region subtag is
-  /// optional.
-  ///
-  /// The values are _case sensitive_, and should match the case of the relevant
-  /// subtags in the [IANA Language Subtag
-  /// Registry](https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry).
-  /// Typically this means the primary language subtag should be lowercase and
-  /// the region subtag should be uppercase.
-  const Locale(this._languageCode, [ this._countryCode ]) : assert(_languageCode != null);
+  /// optional. The values are _case sensitive_, and must match the values
+  /// listed in
+  /// http://unicode.org/repos/cldr/tags/latest/common/validity/language.xml and
+  /// http://unicode.org/repos/cldr/tags/latest/common/validity/region.xml.
+  ///
+  /// This method only produces standards-compliant instances if valid language
+  /// and country codes are provided. Deprecated subtags will be replaced, but
+  /// incorrectly cased strings are not corrected.
+  const Locale(
+    this._languageCode, [
+    this._countryCode,
+  ]) : assert(_languageCode != null),
+       this._scriptCode = null,
+       this._variants = null,
+       this._extensions = null;
+
+  /// Creates a new Locale object with the specified parts.
+  ///
+  /// This is for internal use only. All fields must already be normalized, must
+  /// already be canonicalized. This method does not modify parameters in any
+  /// way or do any syntax checking.
+  ///
+  /// * language, script and region must be in canonical form.
+  /// * iterating over variants must provide variants in alphabetical order.
+  /// * The extensions map must contain only valid key/value pairs. "u" and "t"
+  ///   keys must be present, with an empty string as value, if there are any
+  ///   subtags for those singletons.
+  Locale._internal(
+    String language, {
+    String script,
+    String region,
+    collection.SplayTreeSet<String> variants,
+    collection.SplayTreeMap<String, String> extensions,
+  }) : assert(language != null),
+       assert(language.length >= 2),
+       assert(language.length <= 8),
+       assert(language.length != 4),
+       assert((script ?? 'Xxxx').length == 4),
+       assert((region ?? 'XX').length >= 2),
+       assert((region ?? 'XX').length <= 3),
```

Comment:
this is better expressed as `assert(script == null || script.length == 4)` et al

...but as above, I wouldn't check these lengths at all unless we have a really good reason to do so.

Reply:
This is basic syntactic checking. Full validation is not recommended as a core library function (checking that each individual subtag against the list of valid country codes for example) because that list gets added to (breaking code with old data when presented with a new region/country).

Reply:
hm, i thought that had been fixed. are you on tip of tree?

Reply:
Actually, this also isn't const constructor code. :) So your suggestion works anyway.

(But I'll also double-check what Dart version I'm using and see if that constant-expression problem has been fixed. Thanks!)